File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/TicketService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.seat.SeatSelectDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.seat.SeatUnselectDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.ticket.TicketSearchDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Seat;
import at.ac.tuwien.sepm.groupphase.backend.entity.Ticket;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotAllowedException;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.SeatNotAvailableException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ValidationException;
import org.springframework.data.domain.Page;

import java.util.List;

public interface TicketService {

    /**
     * Generates an Invoice and stores it on the server.
     * @param id of the ticket
     * @return path of the invoice or null if an error occured
     */
    String generateInvoice(Long id);

    /**
     * Returns the path for an invoice
     * @param id of the ticket
     * @return path of the invoice
     */
    String getInvoicePath(Long id);

    /**
     * Returns the path for a cancellation invoice
     * @param id of the ticket
     * @return path of the invoice
     */
    String getCancellationInvoicePath(Long id);

    /**
     * Searches for a Ticket in the Database
     * @param id of the ticket
     * @return the ticket found
     * @throws at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException
     */
    Ticket find(long id) throws NotFoundException;

    /**
     * returns the tickets with the given parameter
     * @param ticketSearchDto search query with values status and user id
     * @return Page with Array of Tickets with the given parameter
     */
    Page<Ticket> filter(TicketSearchDto ticketSearchDto);

    /** @param id of the performance of which we should get or create a new ticket
     * @return the TicketQueueDto which belongs to the current user to the given performance
     * @throws NotFoundException if no performance exists with the given id
     * @throws ValidationException if the requested performance is in the past
     */
    Ticket getOrCreateQueuedTicket(Long id) throws NotFoundException, ValidationException;

    /**
     * Try to select a specific seat or a seat inside an area
     *
     * @param seatSelectDto infos about which seat should be selected
     * @throws SeatNotAvailableException if the requested seat is already taken
     * @throws NotFoundException if the request seat or area does not exist
     * @throws NotAllowedException if the user is not allowed to select seats for the specified ticket
     * @throws ValidationException if the given info for the seat to select is invalid
     * */
    void selectSeat(SeatSelectDto seatSelectDto) throws SeatNotAvailableException, NotFoundException, ValidationException, NotAllowedException;

    /**
     * Try to unselect a specific seat or area
     *
     * @param seatUnselectDto infos about which seat should be selected
     * @throws NotFoundException if the request seat or area does not exist or wasn't selected at first
     * @throws ValidationException if the given info for the seat to unselect is invalid
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * */
    void unselectSeat(SeatUnselectDto seatUnselectDto) throws NotFoundException, ValidationException, NotAllowedException;

    /**
     * Get information about a ticket for checkout
     *
     * @param id of the ticket to get information about
     * @throws NotFoundException if the request ticket with id does not exist
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the performance of the ticket is in the past
     * */
    Ticket getCheckoutInfo(Long id) throws NotFoundException, NotAllowedException, ValidationException;

    /**
     * Buys a ticket by changing its status
     *
     * @param id of the ticket to buy
     * @throws NotFoundException if the request ticket with id does not exist
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the given ticket is not reserved nor in queue
     * */
    void buyTicket(Long id) throws NotFoundException, NotAllowedException, ValidationException;

    /**
     * Reserves a ticket by changing its status
     *
     * @param id of the ticket to reserve
     * @throws NotFoundException if the request ticket with id does not exist
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the given ticket is not in queue
     * */
    void reserveTicket(Long id) throws NotFoundException, NotAllowedException, ValidationException;

    /**
     * Get all selected seats from ticket
     *
     * @param id the ticket from which selected seats should be taken
     * @return List of seats which are selected
     * @throws NotFoundException if there is no ticket with given id
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the Ticket is not valid
     * */
    List<Seat> getSelectedSeats(Long id) throws NotFoundException, NotAllowedException, ValidationException;

    /**
     * Cancel reservation for given ticket
     *
     * @param id of the ticket to cancel
     * @throws NotFoundException if there is no ticket with given id
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the given ticket is not reserved
     * */
    void cancelReservedTicket(Long id) throws NotFoundException, NotAllowedException, ValidationException;

    /**
     * generates the file containing the cancellation invoice
     * @param id the ticket id
     */
    void generateCancellationInvoice(Long id);

    /**
     * Cancel bought ticket
     *
     * @param id of the ticket to cancel
     * @throws NotFoundException if there is no ticket with given id
     * @throws NotAllowedException if the current user is not allowed to access the ticket
     * @throws ValidationException if the given ticket is not bought
     * */
    void cancelBoughtTicket(Long id) throws NotFoundException, NotAllowedException, ValidationException;
}

File: frontend/src/app/services/ticket.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient, HttpParams } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { Globals } from '../global/globals';
import { TicketPreview } from '../dtos/ticket/ticket-preview';
import { TicketPage } from '../dtos/ticket/ticket-page';
import { TicketQueueDto } from "../dtos/ticket/ticket-queue-dto";
import { TicketCheckoutDto } from "../dtos/ticket/ticket-checkout-dto";
import { SeatSelectDto } from "../dtos/seat/seat-select-dto";
import { SeatUnselectDto } from "../dtos/seat/seat-unselect-dto";
import { TicketSeatDto } from "../dtos/ticket/ticket-seat-dto";


@Injectable({
  providedIn: 'root'
})
export class TicketService {

  private messageBaseUri: string = this.globals.backendUri + '/ticket/';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  public getTicket(id: number): Observable<TicketPreview> {
    return this.httpClient.get<TicketPreview>(this.globals.ticketUri + '/' + id);
  }

  public getInvoice(id: number): Observable<Blob> {
    return this.httpClient.get(this.globals.ticketUri + '/' + id + '/invoice', { responseType: 'blob' });
  }

  public getCancellationInvoice(id: number): Observable<Blob> {
    return this.httpClient.get(this.globals.ticketUri + '/' + id + '/storno', { responseType: 'blob' });
  }

  //TODO rename
  /**
   * Loads all tickets from active user
   */
  getFilteredTickets(ticketQuery): Observable<TicketPage> {

    let params = new HttpParams();
    for(let key in ticketQuery) {
      if(ticketQuery[key] !== null) {
        params = params.set(key, ticketQuery[key]);
      }
    }

    return this.httpClient.get<TicketPage>(this.messageBaseUri, {params: params});
  }
  /**
   * Gets a ticket which is alread queued for the current user or creates a new one
   *
   * @param id of the performance for which to search or create a new ticket
   * @return TicketQueueDto with the information about the queued ticket
   * */
  public getOrCreateTicket(id: number): Observable<TicketQueueDto> {
    return this.httpClient.post<TicketQueueDto>(this.globals.ticketUri + "/" + id + "/queue", {});
  }

  /**
   * Get information about a ticket for checkout
   *
   * @param id of the ticket to get
   * @return TicketCheckoutDto with information about the ticket
   * */
  public getCheckoutTicketInfo(id: number): Observable<TicketCheckoutDto> {
    return this.httpClient.get<TicketCheckoutDto>(this.globals.ticketUri + "/" + id + "/checkout")
      .pipe<TicketCheckoutDto>(map<TicketCheckoutDto, TicketCheckoutDto>(r => {
        r.date = new Date(r.date);
        return r
      }));
  }

  /**
   * Get all selected seats for the ticket
   *
   * @param id of the ticket
   * @return Array of TicketSeatDto with selected seats
   * */
  public getSelectedSeats(id: number): Observable<TicketSeatDto[]> {
    return this.httpClient.get<TicketSeatDto[]>(this.globals.ticketUri + "/" + id + "/seats");
  }

  /**
   * Select a specific seat or area
   *
   * @param seatSelectDto infos about which seat to select for which ticket
   * */
  public selectSeat(seatSelectDto: SeatSelectDto): Observable<any> {
    return this.httpClient.post(this.globals.ticketUri + "/select/", {
      ticketId: seatSelectDto.ticketId,
      type: seatSelectDto.type,
      row: seatSelectDto.row,
      col: seatSelectDto.col,
      areaId: seatSelectDto.areaId,
      seatCount: seatSelectDto.seatCount
    });
  }

  /**
   * Unselect a specific seat or area
   *
   * @param seatUnselectDto infos about which seat to unselect for which ticket
   * */
  public unselectSeat(seatUnselectDto: SeatUnselectDto): Observable<any> {
    return this.httpClient.post(this.globals.ticketUri + "/unselect/", {
      ticketId: seatUnselectDto.ticketId,
      type: seatUnselectDto.type,
      row: seatUnselectDto.row,
      col: seatUnselectDto.col,
      areaId: seatUnselectDto.areaId
    });
  }

  /**
   * Buy a ticket
   *
   * @param id of the ticket to buy
   * */
  public buyTicket(id: number): Observable<any> {
    return this.httpClient.patch(this.globals.ticketUri + "/" + id + "/buy", {});
  }

  /**
   * Reserve a ticket
   *
   * @param id of the ticket to reserve
   * */
  public reserveTicket(id: number): Observable<any> {
    return this.httpClient.patch(this.globals.ticketUri + "/" + id + "/reserve", {});
  }

  /**
   * Cancel a ticket reservation
   *
   * @param id of the ticket to cancel
   * */
  public cancelReservedTicket(id: number): Observable<any> {
    return this.httpClient.post(this.globals.ticketUri + "/" + id + "/cancelReserved", {});
  }

  /**
   * Cancel a bought ticket
   *
   * @param id of the ticket to cancel
   * */
  public cancelBoughtTicket(id: number): Observable<any> {
    return this.httpClient.post(this.globals.ticketUri + "/" + id + "/cancelBought", {});
  }

}
